zcash_primitives/src/block.rs

```diff
@@ -0,0 +1,209 @@
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use hex;
+use std::fmt;
+use std::io::{self, Read, Write};
+use std::ops::Deref;
+
+use serialize::Vector;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct BlockHash(pub [u8; 32]);
+
+impl fmt::Display for BlockHash {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        let mut data = self.0.clone();
+        data.reverse();
+        formatter.write_str(&hex::encode(data))
+    }
+}
+
+/// A Zcash block header.
+pub struct BlockHeader(BlockHeaderData);
+
+impl Deref for BlockHeader {
+    type Target = BlockHeaderData;
+
+    fn deref(&self) -> &BlockHeaderData {
+        &self.0
+    }
+}
+
+pub struct BlockHeaderData {
+    pub version: i32,
+    pub prev_block: BlockHash,
+    pub merkle_root: [u8; 32],
+    pub final_sapling_root: [u8; 32],
```

Comment:
Not sure if this is out of scope for the PR, but the `BlockHash` is a typed wrapper for a `[u8; 32]`; are the `merkle_root` and `final_sapling_root` comparable 32-byte arrays, or semantically-different types? Should they also be newtype wrappers?

Reply:
They are all hash outputs, but of at least two different hash functions. In `zcashd` they are represented by the same 256-bit binary blob type, which has the "feature" that when rendered in hex in `zcashd` RPC output, the bytes are reversed (behaviour inherited from Bitcoin Core).

I use separate types for block hashes and transaction IDs because those are the only two that external code is likely to actually interact with, whereas other block header fields (thus far) will only be used internally. Once the refactor is further along and we get to the stage of cleaning up the crate APIs, we should revisit this.

Reply:
Opened https://github.com/zcash/librustzcash/issues/73.